velox/exec/tests/HashJoinTest.cpp

```diff
@@ -150,95 +312,199 @@ class HashJoinTest : public HiveConnectorTestBase {
       "  WHERE t_k0 = u_k0 AND t_k1 = u_k1");
 }
 
-TEST_F(HashJoinTest, normalizedKeyOverflow) {
+TEST_P(MultiThreadedHashJoinTest, normalizedKeyOverflow) {
   testJoin(
       {BIGINT(), VARCHAR(), BIGINT(), BIGINT(), BIGINT(), BIGINT()},
-      1,
-      16000,
-      15000,
+      numDrivers_,
+      1600,
+      3,
+      1500,
+      3,
       "SELECT t_k0, t_k1, t_k2, t_k3, t_k4, t_k5, t_data, u_k0, u_k1, u_k2, u_k3, u_k4, u_k5, u_data FROM "
       "  t, u "
       "  WHERE t_k0 = u_k0 AND t_k1 = u_k1 AND t_k2 = u_k2 AND t_k3 = u_k3 AND t_k4 = u_k4 AND t_k5 = u_k5  ");
 }
 
-TEST_F(HashJoinTest, allTypes) {
+TEST_P(MultiThreadedHashJoinTest, allTypes) {
   testJoin(
       {BIGINT(), VARCHAR(), REAL(), DOUBLE(), INTEGER(), SMALLINT(), TINYINT()},
-      1,
-      16000,
-      15000,
+      numDrivers_,
+      1600,
+      3,
+      1500,
+      3,
       "SELECT t_k0, t_k1, t_k2, t_k3, t_k4, t_k5, t_k6, t_data, u_k0, u_k1, u_k2, u_k3, u_k4, u_k5, u_k6, u_data FROM "
       "  t, u "
       "  WHERE t_k0 = u_k0 AND t_k1 = u_k1 AND t_k2 = u_k2 AND t_k3 = u_k3 AND t_k4 = u_k4 AND t_k5 = u_k5 AND t_k6 = u_k6 ");
 }
 
-TEST_F(HashJoinTest, filter) {
+TEST_P(MultiThreadedHashJoinTest, filter) {
   testJoin(
       {BIGINT()},
-      1,
-      16000,
-      15000,
+      numDrivers_,
+      1600,
+      3,
+      1500,
+      3,
       "SELECT t_k0, t_data, u_k0, u_data FROM "
       "  t, u "
       "  WHERE t_k0 = u_k0 AND ((t_k0 % 100) + (u_k0 % 100)) % 40 < 20",
       "((t_k0 % 100) + (u_k0 % 100)) % 40 < 20");
 }
 
+TEST_P(MultiThreadedHashJoinTest, antiJoinWithNull) {
+  struct {
+    double leftNullRate;
+    double rightNullRate;
+
+    std::string debugString() const {
+      return fmt::format(
+          "leftNullRate: {}, rightNullRate: {}", leftNullRate, rightNullRate);
+    }
+  } testSetttings[] = {
+      {0.0, 1.0}, {0.0, 0.1}, {0.1, 1.0}, {0.1, 0.1}, {1.0, 1.0}, {1.0, 0.1}};
+  for (const auto& testData : testSetttings) {
+    SCOPED_TRACE(testData.debugString());
+
+    auto fuzzerOpts = fuzzerOpts_;
+    fuzzerOpts.nullRatio = testData.leftNullRate;
+    fuzzer_.setOptions(fuzzerOpts);
+    const int numLeftBatches = 10;
+    std::vector<RowVectorPtr> leftBatches;
+    leftBatches.reserve(numLeftBatches);
+    for (int i = 0; i < numLeftBatches; ++i) {
+      leftBatches.push_back(fuzzer_.fuzzRow(leftType_));
+    }
+    // The first half number of build (right) side batches having no nulls to
+    // trigger it later during the processing.
+    const int numRightBatches = 50;
+    fuzzerOpts.vectorSize = 5;
+    fuzzerOpts.nullRatio = 0.0;
+    fuzzer_.setOptions(fuzzerOpts);
+    std::vector<RowVectorPtr> rightBatches;
+    rightBatches.reserve(numRightBatches);
+    int i = 0;
+    for (; i < numRightBatches / 2; ++i) {
+      rightBatches.push_back(fuzzer_.fuzzRow(rightType_));
+    }
+    fuzzerOpts.nullRatio = testData.rightNullRate;
+    fuzzer_.setOptions(fuzzerOpts);
+    for (; i < numRightBatches; ++i) {
```

Comment:
`numRightBatches / 2` ?

Reply:
Not actually, i is continually increased across loops